README.md
# angular gauge

[![Build Status](https://github.com/mattlewis92/angular-gauge/actions/workflows/ci.yml/badge.svg)](https://github.com/mattlewis92/angular-gauge/actions/workflows/ci.yml)
[![codecov](https://codecov.io/gh/mattlewis92/angular-gauge/branch/main/graph/badge.svg)](https://codecov.io/gh/mattlewis92/angular-gauge)
[![npm version](https://badge.fury.io/js/angular-gauge.svg)](http://badge.fury.io/js/angular-gauge)
[![devDependency Status](https://david-dm.org/mattlewis92/angular-gauge/dev-status.svg)](https://david-dm.org/mattlewis92/angular-gauge?type=dev)
[![GitHub issues](https://img.shields.io/github/issues/mattlewis92/angular-gauge.svg)](https://github.com/mattlewis92/angular-gauge/issues)
[![GitHub stars](https://img.shields.io/github/stars/mattlewis92/angular-gauge.svg)](https://github.com/mattlewis92/angular-gauge/stargazers)
[![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/mattlewis92/angular-gauge/main/LICENSE)

## Demo

https://mattlewis92.github.io/angular-gauge/

## Table of contents

- [About](#about)
- [Compatibility](#compatibility)
- [Installation](#installation)
- [Documentation](#documentation)
- [Development](#development)
- [License](#license)

## About

A tiny angular 15.0+ wrapper for https://github.com/naikus/svg-gauge

## Compatibility

| angular-gauge | Angular |
| ------------- | ------- |
| <=3.x.x       | <=9     |
| 4.x.x         | 10      |
| 5.x.x         | 15+     |

## Installation

Install through npm:

```
npm install --save angular-gauge
```

Then include in your apps module:

```typescript
import { Component, NgModule } from '@angular/core';
import { GaugeModule } from 'angular-gauge';

@NgModule({
  imports: [GaugeModule.forRoot()],
})
export class MyModule {}
```

By default no styling is included with the gauge so you need to define some global styles:

```css
mwl-gauge {
  width: 150px;
  height: 150px;
  display: block;
  padding: 10px;
}
mwl-gauge > .gauge > .dial {
  stroke: #eee;
  stroke-width: 2;
  fill: rgba(0, 0, 0, 0);
}
mwl-gauge > .gauge > .value {
  stroke: rgb(47, 227, 255);
  stroke-width: 2;
  fill: rgba(0, 0, 0, 0);
}
mwl-gauge > .gauge > .value-text {
  fill: rgb(47, 227, 255);
  font-family: sans-serif;
  font-weight: bold;
  font-size: 1em;
}
```

Finally use in one of your apps components:

```typescript
import { Component } from '@angular/core';

@Component({
  template: `
    <mwl-gauge
      [max]="100"
      [dialStartAngle]="-90"
      [dialEndAngle]="-90.001"
      [value]="50"
      [animated]="true"
      [animationDuration]="1"
    >
    </mwl-gauge>
  `,
})
export class MyComponent {}
```

You may also find it useful to view the [demo source](https://github.com/mattlewis92/angular-gauge/blob/main/demo/demo.component.ts).

This lib is just a tiny wrapper over the excellent [svg-gauge](https://github.com/naikus/svg-gauge) library.

### Usage without a module bundler

```
<script src="node_modules/angular-gauge/bundles/angular-gauge.umd.js"></script>
<script>
    // everything is exported `angular-gauge` namespace
</script>
```

## Documentation

All documentation is auto-generated from the source via [compodoc](https://compodoc.github.io/compodoc/) and can be viewed here:
https://mattlewis92.github.io/angular-gauge/docs/

## Development

### Prepare your environment

- Install [Node.js](http://nodejs.org/) and NPM (should come with)
- Install local dev dependencies: `npm install` while current directory is this repo

### Development server

Run `npm start` to start a development server on port 8000 with auto reload + tests.

### Testing

Run `npm test` to run tests once or `npm run test:watch` to continually run tests.

### Release

```bash
npm run release
```

## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [5.0.0](https://github.com/mattlewis92/angular-gauge/compare/v4.0.0...v5.0.0) (2023-05-16)

### ⚠ BREAKING CHANGES

- the minimum supported version of angular by this package is now v15

### Features

- upgrade to angular 15/16 ([#27](https://github.com/mattlewis92/angular-gauge/issues/27)) ([203a5f6](https://github.com/mattlewis92/angular-gauge/commit/203a5f60267714fbfe7f035d3d5607ca6f415912))

## [4.0.0](https://github.com/mattlewis92/angular-gauge/compare/v3.1.3...v4.0.0) (2020-10-12)

### ⚠ BREAKING CHANGES

- angular 10 or higher is now required to use this package

### Features

- upgrade to angular 10 ([4736149](https://github.com/mattlewis92/angular-gauge/commit/47361498935bfed9e5f5f91fc1c44a6291d352ec))

<a name="3.1.3"></a>

## [3.1.3](https://github.com/mattlewis92/angular-gauge/compare/v3.1.2...v3.1.3) (2020-07-01)

### Bug Fixes

- allow newer versions of angular in peer dependencies ([f219eb8](https://github.com/mattlewis92/angular-gauge/commit/f219eb8))

<a name="3.1.2"></a>

## [3.1.2](https://github.com/mattlewis92/angular-gauge/compare/v3.1.1...v3.1.2) (2019-08-06)

### Bug Fixes

- allow angular 7 and 8 peer dependency ([869eee0](https://github.com/mattlewis92/angular-gauge/commit/869eee0))
- upgrade svg gauge ([2d56240](https://github.com/mattlewis92/angular-gauge/commit/2d56240))

<a name="3.1.1"></a>

## [3.1.1](https://github.com/mattlewis92/angular-gauge/compare/v3.1.0...v3.1.1) (2018-05-05)

### Bug Fixes

- allow angular 6 peer dependency ([2ff6b49](https://github.com/mattlewis92/angular-gauge/commit/2ff6b49))

<a name="3.1.0"></a>

# [3.1.0](https://github.com/mattlewis92/angular-gauge/compare/v3.0.0...v3.1.0) (2018-01-19)

### Features

- support dynamic value coloring ([2d2408c](https://github.com/mattlewis92/angular-gauge/commit/2d2408c))

<a name="3.0.0"></a>

# [3.0.0](https://github.com/mattlewis92/angular-gauge/compare/v2.0.1...v3.0.0) (2017-12-23)

### Features

- upgrade svg-gauge ([f042c4e](https://github.com/mattlewis92/angular-gauge/commit/f042c4e)), closes [#9](https://github.com/mattlewis92/angular-gauge/issues/9)
- upgrade to angular 5 ([ea9206d](https://github.com/mattlewis92/angular-gauge/commit/ea9206d))
- use ng-packagr for building dist files ([92e6447](https://github.com/mattlewis92/angular-gauge/commit/92e6447)), closes [#10](https://github.com/mattlewis92/angular-gauge/issues/10)

### BREAKING CHANGES

- The svg-gauge dependency was updated, your [CSS will need to be slightly adjusted](https://github.com/naikus/svg-gauge#migration-from-102).
- The svg-gauge package is no longer bundled with this library, system.js users will
  have to add a reference to it in their system.js config
- The `valueTextClass` input was renamed to `valueClass` and the `radius` input was
  renamed to `dialRadius`
- angular 5 or higher is now required to use this package

<a name="2.0.1"></a>

## [2.0.1](https://github.com/mattlewis92/angular-gauge/compare/v2.0.0...v2.0.1) (2017-12-10)

### Bug Fixes

- allow angular 5 peer dependency ([90c6b93](https://github.com/mattlewis92/angular-gauge/commit/90c6b93)), closes [#8](https://github.com/mattlewis92/angular-gauge/issues/8)
- only pass defined options to svg gauge ([f929c40](https://github.com/mattlewis92/angular-gauge/commit/f929c40)), closes [#7](https://github.com/mattlewis92/angular-gauge/issues/7)
- pin svg-gauge version to last version without breaking changes ([e2c0cae](https://github.com/mattlewis92/angular-gauge/commit/e2c0cae))

<a name="2.0.0"></a>

# [2.0.0](https://github.com/mattlewis92/angular-gauge/compare/v1.0.0...v2.0.0) (2017-04-15)

### Features

- **defaults:** allow the default options to be configured in the modules forRoot method ([b10460c](https://github.com/mattlewis92/angular-gauge/commit/b10460c)), closes [#1](https://github.com/mattlewis92/angular-gauge/issues/1)

### BREAKING CHANGES

- **defaults:** angular 4 or higher is now required to run this module. The modules forRoot method
  must now be called when importing it

<a name="1.0.0"></a>

# 1.0.0 (2017-03-22)

### Features

- initial implementation ([c731cee](https://github.com/mattlewis92/angular-gauge/commit/c731cee))


demo/demo.component.ts
import { Component, ViewEncapsulation, OnInit, OnDestroy } from '@angular/core';

@Component({
  selector: 'mwl-demo-app',
  encapsulation: ViewEncapsulation.None, // tslint:disable-line
  template: `
    <div class="gauges-container">
      <mwl-gauge
        [max]="100"
        [dialStartAngle]="-90"
        [dialEndAngle]="-90.001"
        [label]="percentageValue"
        [value]="gaugeValues[1]"
        [animated]="true"
        [animationDuration]="1"
      >
      </mwl-gauge>

      <mwl-gauge
        class="two"
        [max]="100"
        [dialStartAngle]="180"
        [dialEndAngle]="0"
        [value]="gaugeValues[2]"
        [animated]="true"
        [animationDuration]="2"
      >
      </mwl-gauge>

      <mwl-gauge
        class="three"
        [max]="100"
        [value]="gaugeValues[3]"
        [animated]="true"
        [animationDuration]="1.5"
      >
      </mwl-gauge>

      <mwl-gauge
        class="four"
        [max]="100"
        [dialStartAngle]="180"
        [dialEndAngle]="-90"
        [value]="gaugeValues[4]"
        [animated]="true"
        [animationDuration]="2"
      >
      </mwl-gauge>

      <mwl-gauge
        class="five"
        [max]="200"
        [dialStartAngle]="0"
        [dialEndAngle]="-180"
        [value]="gaugeValues[5]"
        [animated]="true"
        [animationDuration]="1"
      >
      </mwl-gauge>

      <mwl-gauge
        class="six"
        [max]="100"
        [dialStartAngle]="90.01"
        [dialEndAngle]="89.99"
        [dialRadius]="10"
        [showValue]="false"
        [value]="gaugeValues[6]"
        [animated]="true"
        [animationDuration]="1"
      >
      </mwl-gauge>

      <mwl-gauge
        class="seven"
        [max]="100"
        [dialStartAngle]="-90"
        [dialEndAngle]="-90.001"
        [showValue]="false"
        [value]="gaugeValues[6]"
        [animated]="true"
        [animationDuration]="1"
      >
      </mwl-gauge>
    </div>
  `,
  styles: [
    `
      body {
        background-color: rgba(0, 0, 0, 0.8);
        color: #999;
        font-family: Hevletica, sans-serif;
      }

      .gauges-container {
        display: flex;
      }

      /* ------ Default Style ---------- */
      mwl-gauge {
        flex: 1;
        display: block;
        padding: 10px;
        background-color: #222;
        margin: 7px;
        border-radius: 3px;
      }
      mwl-gauge > .gauge > .dial {
        stroke: #334455;
        stroke-width: 2;
        fill: rgba(0, 0, 0, 0);
      }
      mwl-gauge > .gauge > .value {
        stroke: rgb(47, 227, 255);
        stroke-width: 2;
        fill: rgba(0, 0, 0, 0);
      }
      mwl-gauge > .gauge > .value-text {
        fill: rgb(47, 227, 255);
        font-family: sans-serif;
        font-weight: bold;
        font-size: 0.8em;
      }

      /* ------- Alternate Style ------- */
      mwl-gauge.two {
      }
      mwl-gauge.two > .gauge > .dial {
        stroke: #334455;
        stroke-width: 10;
      }
      mwl-gauge.two > .gauge > .value {
        stroke: orange;
        stroke-dasharray: none;
        stroke-width: 13;
      }
      mwl-gauge.two > .gauge > .value-text {
        fill: orange;
      }

      /* ------- Alternate Style ------- */
      mwl-gauge.three {
      }
      mwl-gauge.three > .gauge > .dial {
        stroke: #334455;
        stroke-width: 2;
      }
      mwl-gauge.three > .gauge > .value {
        stroke: #c9de3c;
        stroke-width: 5;
      }
      mwl-gauge.three > .gauge > .value-text {
        fill: #c9de3c;
      }

      /* ----- Alternate Style ----- */
      mwl-gauge.four > .gauge > .dial {
        stroke: #334455;
        stroke-width: 5;
      }
      mwl-gauge.four > .gauge > .value {
        stroke: #be80ff;
        stroke-dasharray: none;
        stroke-width: 5;
      }
      mwl-gauge.four > .gauge > .value-text {
        fill: #be80ff;
      }

      /* ----- Alternate Style ----- */
      mwl-gauge.five > .gauge > .dial {
        stroke: #334455;
        stroke-width: 5;
      }
      mwl-gauge.five > .gauge > .value {
        stroke: #f8774b;
        stroke-dasharray: 25 1;
        stroke-width: 5;
      }
      mwl-gauge.five > .gauge > .value-text {
        fill: #f8774b;
        font-size: 0.7em;
      }

      /* ----- Alternate Style ----- */
      mwl-gauge.six > .gauge > .dial {
        stroke: #334455;
        fill: #334455;
        stroke-width: 20;
      }
      mwl-gauge.six > .gauge > .value {
        stroke: #ff6daf;
        stroke-width: 20;
      }
      mwl-gauge.six > .gauge > .value-text {
        fill: #ff6daf;
        font-size: 0.7em;
      }

      mwl-gauge.seven > .gauge > .dial {
        stroke: transparent;
        stroke-width: 5;
        transform: scale(0.9, 0.9) translate3d(5.5px, 5.5px, 0);
        fill: rgba(148, 112, 57, 0.42);
      }
      mwl-gauge.seven > .gauge > .value {
        stroke: #f8774b;
        stroke-dasharray: none;
        stroke-width: 5;
      }
    `,
  ],
})
export class DemoComponent implements OnInit, OnDestroy {
  percentageValue: (value: number) => string;

  gaugeValues: any = {
    1: 100,
    2: 50,
    3: 50,
    4: 50,
    5: 50,
    6: 50,
    7: 50,
  };

  interval: any;

  constructor() {
    this.percentageValue = function (value: number): string {
      return `${Math.round(value)} / ${this['max']}`;
    };
  }

  ngOnInit(): void {
    const updateValues = (): void => {
      this.gaugeValues = {
        1: Math.round(Math.random() * 100),
        2: Math.round(Math.random() * 100),
        3: Math.round(Math.random() * 100),
        4: Math.round(Math.random() * 100),
        5: Math.round(Math.random() * 200),
        6: Math.round(Math.random() * 100),
        7: Math.round(Math.random() * 100),
      };
    };

    const INTERVAL: number = 5000;

    this.interval = setInterval(updateValues, INTERVAL);
    updateValues();
  }

  ngOnDestroy(): void {
    clearInterval(this.interval);
  }
}
